crates/sui-core/src/histogram.rs

```diff
@@ -0,0 +1,402 @@
+        Self { channel: sender }
+    }
+
+    pub fn with_label_values(&self, labels: &[&str]) -> Histogram {
+        let labels = labels.iter().map(ToString::to_string).collect();
+        let labels = HistogramLabelsInner::new(labels);
+        Histogram {
+            labels,
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+impl HistogramLabelsInner {
+    pub fn new(labels: Vec<String>) -> HistogramLabels {
+        // Not a crypto hash
+        let mut hasher = DefaultHasher::new();
+        labels.hash(&mut hasher);
+        let hash = hasher.finish();
+        Arc::new(Self { labels, hash })
+    }
+}
+
+impl PartialEq for HistogramLabelsInner {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
+    }
+}
+
+impl Eq for HistogramLabelsInner {}
+
+impl Hash for HistogramLabelsInner {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.hash.hash(state)
+    }
+}
+
+impl Histogram {
+    pub fn report(&self, v: Point) {
+        match self.channel.try_send((self.labels.clone(), v)) {
+            Ok(()) => {}
+            Err(TrySendError::Closed(_)) => {
+                unreachable!("Histogram reporting task do not stop when Histogram instance exists")
+            }
+            Err(TrySendError::Full(_)) => error!("Histogram channel is full, dropping data"),
+        }
+    }
+
+    pub fn start_timer(&self) -> HistogramTimerGuard {
+        HistogramTimerGuard {
+            histogram: self,
+            start: Instant::now(),
+        }
+    }
+}
+
+impl HistogramCollector {
+    pub async fn run(mut self) {
+        let mut deadline = Instant::now();
+        loop {
+            // We calculate deadline here instead of just using sleep inside cycle to avoid accumulating error
+            deadline += Duration::from_secs(1);
+            if self.cycle(deadline).await.is_err() {
+                return;
+            }
+        }
+    }
+
+    async fn cycle(&mut self, deadline: Instant) -> Result<(), ()> {
+        let mut labeled_data: HashMap<HistogramLabels, Vec<Point>> = HashMap::new();
+        let mut count = 0usize;
+        let mut timeout = tokio::time::sleep_until(deadline).boxed();
+        const MAX_POINTS: usize = 100_000;
+        loop {
+            tokio::select! {
+             _ = &mut timeout => {
+                break;
+            },
+                point = self.channel.recv() => {
+                    count += 1;
+                    if count > MAX_POINTS {
+                        continue;
+                    }
+                    if let Some((label, point)) = point {
+                        let values = labeled_data.entry(label).or_default();
+                        values.push(point);
+                    } else {
+                        // Histogram no longer exists
+                        return Err(());
+                    }
+                }
+                }
+        }
+        if count > MAX_POINTS {
```

Comment:
We should consider using a rotating or ring buffer here.  Newer data just automatically overwrites older data, which makes for a nice sliding window.  No need to error here, because in the long run we will always overflow ...unless there is some periodic reset thing, which I didn't see

Reply:
Oh never mind, it resets every cycle.